File: packages/sparkling-debug-tool/android/src/main/java/com/tiktok/sparkling/debugtool/SparklingDebugTool.kt
// Copyright (c) 2026 TikTok Pte. Ltd.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
package com.tiktok.sparkling.debugtool

import android.app.Activity
import android.app.AlertDialog
import android.app.Application
import android.content.Context
import android.content.pm.ApplicationInfo
import android.os.Looper
import android.text.InputType
import android.widget.EditText
import android.widget.Toast
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentActivity
import com.lynx.devtool.LynxDevtoolEnv
import com.lynx.service.devtool.LynxDevToolService
import com.lynx.service.log.LynxLogService
import com.lynx.tasm.LynxEnv
import com.lynx.tasm.LynxView
import com.lynx.tasm.service.LynxServiceCenter
import com.tiktok.sparkling.debugtool.console.ConsoleLog
import com.tiktok.sparkling.debugtool.console.ConsoleLogStore
import com.tiktok.sparkling.debugtool.console.LynxConsoleLogDelegate
import com.tiktok.sparkling.debugtool.console.LynxConsoleLogcatBridge
import com.tiktok.sparkling.debugtool.console.LynxConsoleSink
import com.tiktok.sparkling.debugtool.floating.SparklingFloatingBallManager
import com.tiktok.sparkling.debugtool.inspect.GlobalPropsRegistry
import com.tiktok.sparkling.debugtool.inspect.GlobalPropsSnapshot
import com.tiktok.sparkling.debugtool.inspect.MethodInvocation
import com.tiktok.sparkling.debugtool.inspect.MethodInvocationStore
import com.tiktok.sparkling.debugtool.inspect.SparklingDebugAutoWiring
import com.tiktok.sparkling.debugtool.inspector.SparklingInspectorFragment

/**
 * Entry point for the Sparkling debug tool. The host typically only needs to
 * call [init] once during `Application.onCreate`; everything else
 * (Lynx debug flags, JS console capture, GlobalProps tracking, Sparkling
 * Method observation, debugTag entry) is wired up automatically when
 * `enableFloatingBall = true`.
 */
object SparklingDebugTool {
    data class Config(
        /**
         * When `true` the host app is expected to call [attachConsoleSink] for each
         * created `LynxView` so JS console messages are captured. When the
         * debugTag entry is enabled the debug tool wires this up itself via
         * `KitViewManager`, so this flag is mainly here for explicit override.
         */
        val enableJsConsole: Boolean = false,
        val enableInNonDebuggableApp: Boolean = false,
        /** Maximum number of console messages kept in the in-memory ring buffer. */
        val consoleBufferSize: Int = 300,
        /**
         * When `true` SparklingView shows its bottom-left debugTag entry. Tapping
         * the tag opens the unified inspector; turning this on also auto-enables
         * every Lynx debug capability and wires up the
         * console / globalProps / method tracers.
         */
        val enableFloatingBall: Boolean = false,
    )

    /**
     * @deprecated Switches are no longer surfaced in the inspector. The tool
     * applies the all-on configuration whenever the debugTag entry is enabled.
     * The class is kept for binary compatibility with hosts that still call
     * [getFlags] / [setFlags] directly.
     */
    data class DebugFlags(
        val lynxDebugEnabled: Boolean = true,
        val devtoolEnabled: Boolean = true,
        val logboxEnabled: Boolean = true,
        val longPressMenuEnabled: Boolean = true,
    )

    private const val PREFS_NAME = "sparkling_debug_tool"
    private const val KEY_DEV_URL = "dev_url"
    private const val KEY_LYNX_DEBUG = "lynx_debug"
    private const val KEY_DEVTOOL = "devtool"
    private const val KEY_LOGBOX = "logbox"
    private const val KEY_LONG_PRESS = "long_press_menu"

    private var currentConfig = Config()

    @JvmStatic
    fun init(
        application: Application,
        config: Config = Config(),
    ) {
        currentConfig = config
        ConsoleLogStore.setCapacity(config.consoleBufferSize)
        if (!isDebuggableApp(application) && !config.enableInNonDebuggableApp) {
            return
        }
        // Always apply the all-on debug flag set when running under the debug
        // tool: switches are no longer user-visible. Hosts that need a custom
        // subset can still call [setFlags] explicitly.
        applyFlags(DebugFlags())
        SparklingFloatingBallManager.install(application)
        SparklingFloatingBallManager.setEnabled(config.enableFloatingBall)
        if (config.enableFloatingBall || config.enableJsConsole) {
            LynxConsoleLogDelegate.installOnce()
            LynxConsoleLogcatBridge.startOnce()
            SparklingDebugAutoWiring.installOnce()
        }
    }

    /** Show or hide the bottom-left debugTag entry at runtime. */
    @JvmStatic
    fun setFloatingBallEnabled(enabled: Boolean) {
        SparklingFloatingBallManager.setEnabled(enabled)
        if (enabled) {
            applyFlags(DebugFlags())
            LynxConsoleLogDelegate.installOnce()
            LynxConsoleLogcatBridge.startOnce()
            SparklingDebugAutoWiring.installOnce()
        }
    }

    /** Legacy hook kept for compatibility; debugTag clicks use SparklingView's default action. */
    @JvmStatic
    fun setFloatingBallActionHandler(handler: SparklingFloatingBallManager.ActionHandler?) {
        SparklingFloatingBallManager.setActionHandler(handler)
    }

    /**
     * Build a fresh inspector fragment. Equivalent to invoking the debugTag's
     * default tap action; useful for hosts that want to surface the
     * inspector via a menu / button.
     */
    @JvmStatic
    fun createFragment(): DialogFragment = SparklingInspectorFragment.newInstance()

    @JvmStatic
    fun getFlags(context: Context): DebugFlags = resolveFlags(context)

    @JvmStatic
    fun setFlags(
        context: Context,
        flags: DebugFlags,
    ) {
        prefs(context)
            .edit()
            .putBoolean(KEY_LYNX_DEBUG, flags.lynxDebugEnabled)
            .putBoolean(KEY_DEVTOOL, flags.devtoolEnabled)
            .putBoolean(KEY_LOGBOX, flags.logboxEnabled)
            .putBoolean(KEY_LONG_PRESS, flags.longPressMenuEnabled)
            .apply()
        if (isDebuggableApp(context) || currentConfig.enableInNonDebuggableApp) {
            applyFlags(flags)
        }
    }

    @JvmStatic
    fun getDevUrl(
        context: Context,
        fallback: String,
    ): String {
        val stored = prefs(context).getString(KEY_DEV_URL, null)?.trim()
        return if (stored.isNullOrEmpty()) fallback else stored
    }

    @JvmStatic
    fun setDevUrl(
        context: Context,
        url: String,
    ) {
        prefs(context)
            .edit()
            .putString(KEY_DEV_URL, url.trim())
            .apply()
    }

    @JvmStatic
    fun isJsConsoleEnabled(): Boolean = currentConfig.enableJsConsole

    /**
     * Hook the Lynx inspector console delegate so that every `console.*` call from
     * the JS context is captured into the shared [ConsoleLogStore]. The debug
     * tool already wires this up automatically via `KitViewManager` when the
     * debugTag entry is enabled; this entry point is kept for hosts that don't
     * use `KitViewManager`.
     *
     * Returns `true` if the delegate was successfully installed.
     */
    @JvmStatic
    fun attachConsoleSink(lynxView: LynxView): Boolean = LynxConsoleSink.attach(lynxView)

    @JvmStatic
    fun detachConsoleSink(lynxView: LynxView) {
        LynxConsoleSink.detach(lynxView)
    }

    /** Open the unified inspector dialog with the given initial tab. */
    @JvmStatic
    @JvmOverloads
    fun openInspectorPanel(
        activity: FragmentActivity,
        initialTab: SparklingInspectorFragment.Tab = SparklingInspectorFragment.Tab.CONSOLE,
    ) {
        val fm = activity.supportFragmentManager
        if (fm.findFragmentByTag(SparklingInspectorFragment.FRAGMENT_TAG) != null) return
        SparklingInspectorFragment
            .newInstance(initialTab)
            .show(fm, SparklingInspectorFragment.FRAGMENT_TAG)
    }

    /** Open the inspector with the Console tab pre-selected. */
    @JvmStatic
    fun openConsolePanel(activity: FragmentActivity) {
        openInspectorPanel(activity, SparklingInspectorFragment.Tab.CONSOLE)
    }

    /**
     * Open the inspector with the Sparkling Method invocations tab
     * pre-selected.
     */
    @JvmStatic
    fun openMethodInvocationPanel(activity: FragmentActivity) {
        openInspectorPanel(activity, SparklingInspectorFragment.Tab.METHODS)
    }

    /** Open the inspector with the GlobalProps tab pre-selected. */
    @JvmStatic
    fun openGlobalPropsPanel(activity: FragmentActivity) {
        openInspectorPanel(activity, SparklingInspectorFragment.Tab.GLOBAL_PROPS)
    }

    /** Append one method invocation record (already-formatted strings). */
    @JvmStatic
    fun recordMethodInvocation(record: MethodInvocation) {
        MethodInvocationStore.add(record)
    }

    /** Replace an in-flight invocation record (matched by id). */
    @JvmStatic
    fun updateMethodInvocation(record: MethodInvocation) {
        MethodInvocationStore.update(record)
    }

    @JvmStatic
    fun clearMethodInvocations() {
        MethodInvocationStore.clear()
    }

    /**
     * Override the GlobalProps supplier. The debug tool registers a default
     * provider that walks `KitViewManager` reflectively when `init` is called
     * with `enableFloatingBall = true`; hosts only need to call this if they
     * want to expose a different data source.
     */
    @JvmStatic
    fun setGlobalPropsProvider(provider: GlobalPropsRegistry.Provider?) {
        GlobalPropsRegistry.setProvider(provider)
    }

    /**
     * Push one externally-collected console line (e.g. from the host
     * `HybridLogger`) into the shared store. The native panel will display it
     * alongside JS console messages captured via [attachConsoleSink].
     */
    @JvmStatic
    fun recordConsoleLine(
        level: String,
        tag: String,
        message: String,
    ) {
        ConsoleLogStore.add(ConsoleLog.fromPlain(level, tag, message))
    }

    /**
     * Snapshot helper kept for backwards compatibility with previous tooling that
     * rendered the console as a single text dump.
     */
    @JvmStatic
    fun getConsoleLines(): List<String> = ConsoleLogStore.snapshot().map { "[${it.type}] ${if (it.tag.isEmpty()) "" else it.tag + ": "}${it.message}" }

    @JvmStatic
    fun clearConsoleLines() {
        ConsoleLogStore.clear()
    }

    @JvmStatic
    fun showDevUrlDialog(
        activity: Activity,
        initialUrl: String? = null,
        onSaved: ((String) -> Unit)? = null,
    ) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            activity.runOnUiThread { showDevUrlDialog(activity, initialUrl, onSaved) }
            return
        }

        val input =
            EditText(activity).apply {
                setText(initialUrl ?: "")
                hint = "http://127.0.0.1:5969/main.lynx.bundle"
                inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_URI
                setSelection(text.length)
            }

        val dialog =
            AlertDialog
                .Builder(activity)
                .setTitle("Set Sparkling Dev URL")
                .setMessage("Update the main Lynx bundle URL for debug mode.")
                .setView(input)
                .setNegativeButton("Cancel", null)
                .setPositiveButton("Save", null)
                .create()

        dialog.setOnShowListener {
            val saveButton = dialog.getButton(AlertDialog.BUTTON_POSITIVE)
            saveButton.setOnClickListener {
                val value =
                    input.text
                        ?.toString()
                        ?.trim()
                        .orEmpty()
                if (value.isEmpty()) {
                    Toast.makeText(activity, "Dev URL cannot be empty", Toast.LENGTH_SHORT).show()
                    return@setOnClickListener
                }
                setDevUrl(activity, value)
                onSaved?.invoke(value)
                dialog.dismiss()
            }
        }

        dialog.show()
    }

    private fun prefs(context: Context) = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private fun isDebuggableApp(context: Context): Boolean = (context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0

    private fun resolveFlags(context: Context): DebugFlags {
        val prefs = prefs(context)
        return DebugFlags(
            lynxDebugEnabled = prefs.getBoolean(KEY_LYNX_DEBUG, true),
            devtoolEnabled = prefs.getBoolean(KEY_DEVTOOL, true),
            logboxEnabled = prefs.getBoolean(KEY_LOGBOX, true),
            longPressMenuEnabled = prefs.getBoolean(KEY_LONG_PRESS, true),
        )
    }

    private fun applyFlags(flags: DebugFlags) {
        LynxDevToolService.INSTANCE.setLynxDebugPresetValue(flags.lynxDebugEnabled)
        LynxDevToolService.INSTANCE.setLogBoxPresetValue(flags.logboxEnabled)
        LynxDevToolService.INSTANCE.setLoadQJSBridge(flags.devtoolEnabled)

        LynxServiceCenter.inst().registerService(LynxLogService)
        LynxServiceCenter.inst().registerService(LynxDevToolService.INSTANCE)

        LynxEnv.inst().enableLynxDebug(flags.lynxDebugEnabled)
        LynxEnv.inst().enableDevtool(flags.devtoolEnabled)
        LynxEnv.inst().enableLogBox(flags.logboxEnabled)
        LynxDevtoolEnv.inst().enableLongPressMenu(flags.longPressMenuEnabled)
    }
}

File: packages/sparkling-debug-tool/ios/Sources/SparklingDebugTool.swift
// Copyright (c) 2026 TikTok Pte. Ltd.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

import DebugRouter
import Foundation
import Lynx
import UIKit

@objcMembers
public class SparklingDebugTool: NSObject {
    private static let devURLKey = "sparkling.debug.dev_url"

    public static func setup() {
        // Preset values must be set BEFORE LynxEnv flags so the DevTool
        // service picks them up during initialization.

        if let devtool = LynxServices.getInstanceWith(LynxServiceDevToolProtocol.self)
            as? LynxServiceDevToolProtocol
        {
            devtool.logBoxPresetValue = true
            devtool.lynxDebugPresetValue = true
        }

        let env = LynxEnv.sharedInstance()
        env.lynxDebugEnabled = true
        env.devtoolEnabled = true
        env.logBoxEnabled = true

        // Enable DebugRouter so the DevTool desktop app can discover and
        // connect to this app (via USB or network).
        DebugRouter.instance().enableAllSessions()
    }

    public static func devURL(fallback: String) -> String {
        let stored = UserDefaults.standard.string(forKey: devURLKey)?.trimmingCharacters(in: .whitespacesAndNewlines)
        if let stored, !stored.isEmpty {
            return stored
        }
        return fallback
    }

    public static func setDevURL(_ url: String) {
        let normalized = url.trimmingCharacters(in: .whitespacesAndNewlines)
        UserDefaults.standard.set(normalized, forKey: devURLKey)
    }

    /// Attach the JS console sink to the given LynxView so that any
    /// `console.*` payload it emits flows into [SparklingConsoleStore].
    @discardableResult
    public static func attachConsoleSink(to lynxView: LynxView) -> Bool {
        return SparklingLynxConsoleSink.attach(to: lynxView)
    }

    /// Detach the JS console sink previously attached via [attachConsoleSink].
    public static func detachConsoleSink(from lynxView: LynxView) {
        SparklingLynxConsoleSink.detach(from: lynxView)
    }

    /// Show or hide the Sparkling bottom-left debugTag. The tag is hosted in
    /// a dedicated overlay window so it follows the foreground scene.
    /// Enabling the tag also auto-installs the LynxView lifecycle tracker so
    /// the JS console + GlobalProps panels start receiving data without any
    /// host wiring.
    public static func setFloatingBallEnabled(_ enabled: Bool) {
        SparklingFloatingBallManager.shared.setEnabled(enabled)
        if enabled {
            SparklingDebugAutoWiring.installOnce()
            // The default tap presents the unified inspector.
            SparklingFloatingBallManager.shared.setActionHandler(SparklingDefaultFloatingBallHandler.shared)
        }
    }

    /// Whether the debugTag entry is currently enabled.
    public static var isFloatingBallEnabled: Bool {
        SparklingFloatingBallManager.shared.isEnabled
    }

    /// Override the default tap behaviour of the debugTag entry.
    public static func setFloatingBallActionHandler(_ handler: SparklingFloatingBallActionHandler?) {
        SparklingFloatingBallManager.shared.setActionHandler(handler)
    }

    /// Present the unified inspector panel (Console / GlobalProps / Method
    /// invocations) as a half-sheet modal.
    public static func presentInspectorPanel(
        from presenter: UIViewController,
        initialTab: SparklingInspectorViewController.Tab = .console
    ) {
        DispatchQueue.main.async {
            let inspector = SparklingInspectorViewController(initialTab: initialTab)
            Self.applyHalfSheetPresentation(to: inspector)
            presenter.present(inspector, animated: true)
        }
    }

    /// Present the inspector with the Console tab selected.
    public static func presentConsolePanel(from presenter: UIViewController) {
        presentInspectorPanel(from: presenter, initialTab: .console)
    }

    /// Present the inspector with the GlobalProps tab selected.
    public static func presentGlobalPropsPanel(from presenter: UIViewController) {
        presentInspectorPanel(from: presenter, initialTab: .globalProps)
    }

    /// Present the inspector with the Sparkling Method tab selected.
    public static func presentMethodInvocationPanel(from presenter: UIViewController) {
        presentInspectorPanel(from: presenter, initialTab: .methods)
    }

    /// Configure a navigation controller to be presented as a draggable
    /// half-height sheet (medium detent by default, expandable to large).
    /// Falls back to system page sheet on iOS 13/14.
    public static func applyHalfSheetPresentation(to viewController: UIViewController) {
        viewController.modalPresentationStyle = .pageSheet
        if #available(iOS 15.0, *) {
            if let sheet = viewController.sheetPresentationController {
                sheet.detents = [.medium(), .large()]
                sheet.prefersGrabberVisible = true
                sheet.prefersScrollingExpandsWhenScrolledToEdge = false
                sheet.prefersEdgeAttachedInCompactHeight = true
                sheet.widthFollowsPreferredContentSizeWhenEdgeAttached = true
                sheet.preferredCornerRadius = 0
            }
        }
    }

    /// Register a closure that returns one snapshot per live LynxView. The
    /// inspector panel calls this on demand.
    public static func setGlobalPropsProvider(_ provider: (() -> [SparklingGlobalPropsSnapshot])?) {
        SparklingGlobalPropsRegistry.shared.setProvider(provider)
    }

    /// Append a method invocation record (or update an existing one with the
    /// same id) into the in-memory store. The host normally registers a
    /// `SparklingMethodInvocationObserver` that calls these helpers.
    public static func recordMethodInvocation(_ invocation: SparklingMethodInvocation) {
        SparklingMethodInvocationStore.shared.add(invocation)
    }

    public static func updateMethodInvocation(_ invocation: SparklingMethodInvocation) {
        SparklingMethodInvocationStore.shared.update(invocation)
    }

    public static func clearMethodInvocations() {
        SparklingMethodInvocationStore.shared.clear()
    }

    public static func showDevURLDialog(
        from viewController: UIViewController,
        initialURL: String?,
        onSaved: ((String) -> Void)? = nil
    ) {
        DispatchQueue.main.async {
            let alert = UIAlertController(
                title: "Set Sparkling Dev URL",
                message: "Update the main Lynx bundle URL for debug mode.",
                preferredStyle: .alert
            )

            alert.addTextField { textField in
                textField.placeholder = "http://localhost:5969/main.lynx.bundle"
                textField.keyboardType = .URL
                textField.text = initialURL
                textField.clearButtonMode = .whileEditing
            }

            alert.addAction(UIAlertAction(title: "Cancel", style: .cancel))
            alert.addAction(
                UIAlertAction(
                    title: "Save", style: .default,
                    handler: { _ in
                        let value = alert.textFields?.first?.text?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                        guard !value.isEmpty else {
                            return
                        }
                        setDevURL(value)
                        onSaved?(value)
                    }))

            viewController.present(alert, animated: true)
        }
    }
}
